### PR TITLE
refactor(dev-infra): use a singleton for GitClient

### PR DESCRIPTION
**Pull request by josephperrott**
Creates a singleton class for GitClient rather than relying on creating an instance to
require being passed around throughout its usages.

### CODE REVIEW

**Review by gkalpak — COMMENTED** [review]
A few (mostly minor) comments, but overall lgtm :+1:

File: dev-infra/utils/git/index.ts [hunk 1]
```diff
@@ -44,34 +44,76 @@ export class GitCommandError extends Error {
...
+  static getInstance(): GitClient {
+    if (!GitClient.instance) {
+      GitClient.instance = new GitClient();
+    }
+    return GitClient.instance;
+  }
+
   /** Whether verbose logging of Git actions should be used. */
-  static LOG_COMMANDS = true;
+  private verboseLogging = false;
```
[the remark this answers is not in the document]
Comment (josephperrott) [reply dev-infra/utils/git/index.ts:64]: Whoops, I set it to `false` while testing.  Though I do think that `false` should be the default.  But that as you said is for another PR.

File: dev-infra/utils/git/github.ts [hunk 2]
```diff
@@ -26,31 +32,67 @@ export class GithubApiRequestError extends Error {
   }
 }
 
+/** Error for failed Github Graphql equests. */
+export class GithubGraphqlError extends Error {}
```
Comment (devversion) [inline dev-infra/utils/git/github.ts:36]: Nit: Above we use `GraphQL` but here `Graphql`. Should we be consistent? I prefer the latter personally.

File: dev-infra/utils/git/github.ts [hunk 3]
```diff
@@ -26,31 +32,67 @@ export class GithubApiRequestError extends Error {
...
 
     this.hook.error('request', error => {
       // Wrap API errors in a known error class. This allows us to
       // expect Github API errors better and in a non-ambiguous way.
       throw new GithubApiRequestError(error.status, error.message);
     });
 
-    // Create authenticated graphql client.
-    this.graphql = new GithubGraphqlClient(token);
+    this.hook.before('request', async options => {
```
Comment (devversion) [inline dev-infra/utils/git/github.ts:59]: Can you add a comment on why we do this? I guess we do this because we now can have a token set at _any point_ but the API instance is already initialized, so we need to intercept the request.

Comment (devversion) [reply dev-infra/utils/git/github.ts:59]: I think a lot of this could be easier (also for GraphQL) if we just initialize the token at construction (as done before), but have another static method like:

```ts
const git = GitClient.withAuthenticated(token);
```

This could also help us guard that other code blocks accessing the `GitClient` cannot accidentally make changes (unlikely; but just in theory). What do you think?

File: dev-infra/utils/git/github.ts [hunk 4]
```diff
@@ -26,31 +32,67 @@ export class GithubApiRequestError extends Error {
...
+      if (this._token) {
+        options.headers = {...options.headers, authorization: `token ${this._token}`};
+      }
+    });
+    // Note: The prototype must be set explictly as Github's Octokit class is a non-standard class
+    // definition which adjusts the prototype chain.
+    // See:
+    //    https://github.com/Microsoft/TypeScript/wiki/FAQ#why-doesnt-extending-built-ins-like-error-array-and-map-work
+    //    https://github.com/octokit/rest.js/blob/7b51cee4a22b6e52adcdca011f93efdffa5df998/lib/constructor.js
+    Object.setPrototypeOf(this, GithubClient.prototype);
```
Comment (devversion) [inline dev-infra/utils/git/github.ts:69]: Out of curiosity: Why is this needed now? Is something using `instanceof GithubClient` on the `GitClient`?

Comment (josephperrott) [reply dev-infra/utils/git/github.ts:69]: We just never realized it was a problem because we never used `getCurrentUser()` method, so we never realized it wasn't available at runtime.

All of the property values work, its just the `prototype` attributes that fail to be available.

File: dev-infra/utils/git/github.ts (hunk 3 above)
Comment (josephperrott) [reply dev-infra/utils/git/github.ts]: I created https://github.com/angular/angular/pull/41515/commits/2e4955b95f89459e42d035dc140e6894fabd966f to see what it would like it.

Personally, I don't love it because then for authenticated uses, we have to pass the token to wherever the `GitClient` is needed, which ends up kind of negating the value we got from not having to pass the `GitClient` instance around.

---

Another option is for us to have `GitClient` have `getAuthenticatedInstance` and `getInstance` static methods as well as a `buildAuthenticatedInstance` method for setting it initially, then we could have the `addGithubTokenOption` utility function build the authenticated instance so its available for commands which rely on it.

It could be something like:

`github-yargs.ts`
```ts
export function addGithubTokenOption(yargs: Argv): ArgvWithGithubToken {
  return yargs
      // 'github-token' is casted to 'githubToken' to properly set up typings to reflect the key in
      // the Argv object being camelCase rather than kebob case due to the `camel-case-expansion`
      // config: https://github.com/yargs/yargs-parser#camel-case-expansion
      .option('github-token' as 'githubToken', {
        type: 'string',
        description: 'Github token. If not set, token is retrieved from the environment variables.',
        coerce: (token: string) => {
          const githubToken = token || process.env.GITHUB_TOKEN || process.env.TOKEN;
          if (!githubToken) {
            error(red('No Github token set. Please set the `GITHUB_TOKEN` environment variable.'));
            error(red('Alternatively, pass the `--github-token` command line flag.'));
            error(yellow(`You can generate a token here: ${GITHUB_TOKEN_GENERATE_URL}`));
            process.exit(1);
          }
          GitClient.buildAuthenticatedInstance(githubToken);
          return githubToken;
        },
      })
      .default('github-token' as 'githubToken', '', '<LOCAL TOKEN>');
}
```


`git/index.ts`
```ts
export class GitClient {
  private static instance: GitClient;
  private static authenticatedInstance: GitClient;
  
  static getInstance(): GitClient {
    if (!GitClient.instance) {
      GitClient.instance = new GitClient();
    }
    return GitClient.instance;
  }

  static getAuthenticatedInstance(): GitClient {
    if (!GitClient.authenticatedInstance) {
      throw Error('No authenticated instance built yet');
    }
    return GitClient.authenticatedInstance;
  }

  static buildAuthenticatedInstance(token: string) {
    if (GitClient.authenticatedInstance) {
      throw Error('Cannot recreate authenticated instance');
    }
    return GitClient.authenticatedInstance;
  }

...

}
```

**Review by gkalpak — APPROVED** [review]
LGTM (as soon as CI is green :wink:) 🚀

File: dev-infra/utils/git/github.ts (hunk 3 above)
Comment (devversion) [reply dev-infra/utils/git/github.ts]: > I created 2e4955b to see what it would like it.

> Personally, I don't love it because then for authenticated uses, we have to pass the token to wherever the > GitClient is needed, which ends up kind of negating the value we got from not having to pass the GitClient instance around.

Yeah, good point. Similary to the snippet you shared, we could have three methods on the `GitClient`:

```ts
class GitClient {
  static default: this;

  /** Can throw; a getter or a function if not authenticated */
  static authenticated: this;
  /** Authenticates and returns an instance of the Git client; can be later retrieved */
  static authenticateWithToken(token: string): this;
}
```

This allows us to easily provide the key as we do right now, simplify the `GitClient` logic significantly as mentioned above, and also will ensure we explicitly request an authenciated instance (compared to always getting the same one)

File: dev-infra/utils/git/github.ts (hunk 4 above)
Comment (devversion) [reply dev-infra/utils/git/github.ts:67]: Makes sense!

File: dev-infra/utils/git/github.ts (hunk 3 above)
Comment (josephperrott) [reply dev-infra/utils/git/github.ts]: Updated as discussed here, PTAL.

File: dev-infra/utils/git/github.ts [hunk 5]
```diff
@@ -26,20 +32,24 @@ export class GithubApiRequestError extends Error {
...
  * Additionally, provides convenience methods for actions which require multiple requests, or
  * would provide value from memoized style responses.
  **/
 export class GithubClient extends Octokit {
-  /** The Github GraphQL (v4) API. */
-  graphql: GithubGraphqlClient;
-
   /** The current user based on checking against the Github API. */
   private _currentUser: string|null = null;
+  /**  */
```
Comment (devversion) [inline dev-infra/utils/git/github.ts:47]: Nit: Leftover

File: dev-infra/utils/git/github.ts [hunk 6]
```diff
@@ -49,8 +59,33 @@ export class GithubClient extends Octokit {
...
+    }
+
+    /**
+     * Parameters for the graphql request, always setting the authorization header for the request.
+     */
+    const graphqlRequestParams: RequestParameters = {
+      ...params,
+      headers: {
+        authorization: `token ${this.token}`,
+      },
```
Comment (devversion) [inline dev-infra/utils/git/github.ts:86]: Optional: I assume we could now go back to just initializing an authenticated `graphql` instance on construction? (so we don't need to do it on each call and can use `.defaults`)

**Review by devversion — APPROVED** [review]
LGTM 👍

**Comment by zarend** [comment]
FYI @josephperrott this was merged into only master and not patch because of a merge conflict with the patch branch.
